pkg/controller/servicebindingrequest/annotations/annotations.go

```diff
@@ -0,0 +1,72 @@
+package annotations
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+)
+
+// bindingType encodes the medium the binding should deliver the configuration value.
+type bindingType string
+
+const (
+	// BindingTypeVolumeMount indicates the binding should happen through a volume mount.
+	BindingTypeVolumeMount bindingType = "volumemount"
```

Comment:
They're used in https://github.com/isutton/service-binding-operator/blob/71a885b9f2df0bb64c5f4bcb65f24d6c90a298eb/pkg/controller/servicebindingrequest/servicecontext.go#L133-L136.

Reply:
IMHO, reasoning about the type by using string compare could be errorprone and harder to maintain. How about that we make these constants private, and implement on `annotations.Result` a few functions that could give us info what type we are dealing with. For example:

```
if r.IsBindingTypeVolumeMount() {
  volumeKeys = append(volumeKeys, r.Path)
}
```

An alternative would be to define `VolumeKeys()` function on `Result` type, so that we could do something like:

```
volumeKeys = append(volumeKeys, r.VolumeKeys()...)
```

Reply:
> IMHO, reasoning about the type by using string compare could be errorprone and harder to maintain. 

I'm afraid there's a type comparison in there as well. I've simplified it on the example below.

```go
package main

import (
	"fmt"
)

func main() {
	type bindingType string

	s := "foo"
	b := bindingType("foo")

	fmt.Printf("%d", b == s)
}
```

This triggers an error:

```
invalid operation: b == s (mismatched types bindingType and string)
```

> How about that we make these constants private, and implement on `annotations.Result` a few functions that could give us info what type we are dealing with. For example:
> 
> ```
> if r.IsBindingTypeVolumeMount() {
>   volumeKeys = append(volumeKeys, r.Path)
> }
> ```

`Result` was written to be a struct only, not an object. Additionally, this method single purpose is to perform that check, so it seems unworthy for now.

> An alternative would be to define `VolumeKeys()` function on `Result` type, so that we could do something like:
> 
> ```
> volumeKeys = append(volumeKeys, r.VolumeKeys()...)
> ```

I'm not very keen on adding this construct to `Result`, since it has only the data that has been extracted, and it is subject to interpretation of its owner. It adds too much knowledge to that struct.

Either way, this refactoring can wait until a later stage.

Reply:
> Result was written to be a struct only, not an object. Additionally, this method single purpose is to perform that check, so it seems unworthy for now.

I am not speaking about object here - it is a function that hides how the type detection is performed. This gives us more flexibility, and less code refactoring later on. `r.IsBindingTypeVolumeMount()` is just syntactic sugar, we could also have function `IsBindingTypeVolumeMount(r)`. Annotatation package should provide all necessary logic to reason about result type, that should be leaked into another package.

Reply:
Please submit a PR with this change.